### PR TITLE
Convert TeacherHomepage to functional component

### PR DESCRIPTION
**Pull request by maureensturgeon**
I wanted to make the teacher homepage banner dismiss-able but decided I'd convert the component to a functional component first. In the process, I discovered the unit tests were seriously lacking so I added a couple of basic unit tests.

## Testing story
Tested locally that census banner and teacher announcements were still working as expected and the page rendered without errors. Also added to the unit tests.

## Follow-up work
There is a lot of census banner logic that could be extracted to a different component, but there were enough changes here that I didn't want to add that. It's potential follow-up work that we could do later to further clean up this component.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx [hunk 1]
```diff
@@ -20,267 +19,247 @@ import CensusTeacherBanner from '../census2017/CensusTeacherBanner';
...
+  const [
+    censusBannerTeachesSelection,
+    setCensusBannerTeachesSelection
+  ] = useState(null);
+  const [
+    censusBannerInClassSelection,
+    setCensusBannerInClassSelection
+  ] = useState(null);
+  const [showCensusUnknownError, setShowCensusUnknownError] = useState(false);
+  const [showCensusInvalidError, setShowCensusInvalidError] = useState(false);
```
Comment (jamescodeorg) [inline apps/src/templates/studioHomepages/TeacherHomepage.jsx:62]: Does the "verticalness" of our formatting rules bug anyone else?  (Probably not much we can do about it, mostly just wondering if it's just me.)

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx [hunk 2]
```diff
@@ -20,267 +19,247 @@ import CensusTeacherBanner from '../census2017/CensusTeacherBanner';
...
+              onDismiss={() => dismissAndHideCensusBanner()}
+              onPostpone={() => postponeCensusBanner()}
+              onTeachesChange={event =>
+                setCensusBannerTeachesSelection(
+                  event.target.id === 'teachesYes'
+                )
+              }
+              onInClassChange={event =>
+                setCensusBannerInClassSelection(event.target.id === 'inClass')
+              }
```
Comment (jamescodeorg) [inline apps/src/templates/studioHomepages/TeacherHomepage.jsx:213]: Interesting that the bulk of the props for TeacherHomepage are passthrough to CensusTeacherBanner.  I wonder if there's a way to isolate these a bit more as we refactor.

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx [hunk 3]
```diff
@@ -20,267 +19,247 @@ import CensusTeacherBanner from '../census2017/CensusTeacherBanner';
...
-    this.setState({
-      showCensusUnknownError: true
-    });
-  };
-
-  dismissCensusBanner(onSuccess, onFailure) {
+  const dismissCensusBanner = (onSuccess, onFailure) => {
     $.ajax({
-      url: `/api/v1/users/${this.props.teacherId}/dismiss_census_banner`,
+      url: `/api/v1/users/${teacherId}/dismiss_census_banner`,
```
Comment (jamescodeorg) [inline apps/src/templates/studioHomepages/TeacherHomepage.jsx:102]: Oh, interesting that we have an API for this.  I kind of expected this to just be tracked client-side in the cookie or session state.

**Review by jamescodeorg — APPROVED** [review]
If you're interested in another opinion, Maddie might be interested in taking a look too.  Lots of interesting hooks here!

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx (hunk 1 above)
Comment (maureensturgeon) [reply apps/src/templates/studioHomepages/TeacherHomepage.jsx:62]: I wouldn't mind allowing our lines to be a bit longer, this is hard to parse. I could also rename these to be a bit shorter and then it'd be easier to read.

Comment (jamescodeorg) [reply apps/src/templates/studioHomepages/TeacherHomepage.jsx:62]: Eh, no need to rename to workaround linting rules (but feel free to rename if you think different names would improve clarity).

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx (hunk 2 above)
Comment (maureensturgeon) [reply apps/src/templates/studioHomepages/TeacherHomepage.jsx:213]: We could pass them through as one object in the props. For example TeacherHomepage would take a prop called censusData which would include the props that are just being threaded through. Also if we moved some of the census banner specific functionality into the CensusBanner component (where I think it should be since I think that component is only used here), that would also reduce the number of props we're passing in here.

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx (hunk 3 above)
Comment (maureensturgeon) [reply apps/src/templates/studioHomepages/TeacherHomepage.jsx:102]: Maybe there's a requirement that teachers can only submit census information once? This would enforce that. For banners where the user is submitting information, it makes sense we'd want to track whether it should be shown in the DB

File: apps/src/templates/studioHomepages/TeacherHomepage.jsx (hunk 2 above)
Comment (maddiedierker) [reply apps/src/templates/studioHomepages/TeacherHomepage.jsx:213]: i definitely agree that the CensusTeacherBanner logic should be moved there. i think [all of these functions](https://github.com/code-dot-org/code-dot-org/blob/8b73c76f3b5c729af1ce4569e639f8b2bbc9efd6/apps/src/templates/studioHomepages/TeacherHomepage.jsx#L80-L132) (API requests and event handlers that only apply to the census banner) should live there instead.

another way we could handle this would be to accept CensusTeacherBanner as an optional prop in this component:

```javascript
function TeacherHomepage({censusBanner}) {
  return <>
    {/* markup before census banner */}
    {censusBanner}
    {/* markup after census banner */}
  </>;
}

// usage in _homepage.js
<TeacherHomepage
  censusBanner={<CensusTeacherBanner {..allTheCensusProps} />}
/>
```

caveat being that there may be some state-management TeacherHompage is doing that might make this a less ideal solution (e.g., currently, TeacherHomepage cares when displayCensusBanner is false -- should it?)

File: apps/test/unit/templates/studioHomepages/TeacherHomepageTest.js [hunk 4]
```diff
@@ -40,23 +50,69 @@ describe('TeacherHomepage', () => {
...
+      buttonText: 'press me',
+      description: 'description',
+      link: '/link',
+      image: '/image',
+      id: 'id'
+    };
+    const wrapper = setUp({
+      announcement
+    });
+    assert(!wrapper.find('Notification').exists());
```
Comment (maddiedierker) [inline apps/test/unit/templates/studioHomepages/TeacherHomepageTest.js:91]: nit: use `refute` for negative test cases like this

**Review by maddiedierker — APPROVED** [review]
yay hooks! i love being tagged on these so definitely keep doing that 😄 

hook

File: apps/test/unit/templates/studioHomepages/TeacherHomepageTest.js (hunk 4 above)
Comment (maureensturgeon) [reply apps/test/unit/templates/studioHomepages/TeacherHomepageTest.js:91]: I'll make this change in my next PR so I don't have to wait for this branch to build again

Comment (maureensturgeon) [reply apps/test/unit/templates/studioHomepages/TeacherHomepageTest.js:91]: Oh actually I just got the error `TypeError: (0 , _chai.refute) is not a function` are you sure `refute` is the right method here?

Comment (maddiedierker) [reply apps/test/unit/templates/studioHomepages/TeacherHomepageTest.js:91]: you're totally right, i think my brain was in ruby-land 😅 we have a wrapper to create `refute` in [enroll_formTest.js](https://github.com/code-dot-org/code-dot-org/blob/959d57320d48c417f7567d9cd6cdb3685d84cf29/apps/test/unit/code-studio/pd/workshop_enrollment/enroll_formTest.js#L9), but that's unnecessary here. alternatively, we could use `expect(x).not.to.exist` ([chai `exist` doc](https://www.chaijs.com/api/bdd/#method_exist))
